### ISSUE TITLE
Add support for the MLXLM models for the Xgrammar backend

### ISSUE THREAD

**Issue by RobinPicard** [dottxt-ai/outlines#1698]
The `Xgrammar` backend currently relies on a logits precessor directly provided by the library. An issue is that this logits processor only works with `torch` tensors (what's used by the `Transformers` model). As a result, it cannot be used with the `MLXLM` models that use `mlx` tensors while `xgrammar` does provide support for `mlx`.

To fix that issue, we should use the lower-level API of the `xgrammar` library and create a logits processor for that backend ourselves. That way, we would be able to handle both `torch` and `mlx` tensors. This is already what we do for the `LLGuidance` backend, so it should be possible to draw inspiration from it.

Additionally, creating our own logits processor is needed to add extra features such as a thinking mode.
